### ISSUE TITLE
Upgrade Mastodon 3.1.3 (on droplet from Mastodon image in Digital Ocean)

### ISSUE THREAD

**Issue by olivievranska** [mastodon/mastodon#20128]
### Steps to reproduce the problem

I created droplet from Mastodon image in Digital Ocean. There is installed 3.1.3 version of Mastodon.

I followed "upgrading to a new release" guide in https://docs.joinmastodon.org/admin/upgrading/, I got mimemagic error and I cannot continue.

1. su - mastodon
2. cd /home/mastodon/live
3. git fetch --tags
4. git checkout v3.1.4 (the first higher)
mastodon@cztwitter:~/live$ git checkout v3.1.4
Previous HEAD position was 85f0e0a60 Update CHANGELOG.md (#13397)
HEAD is now at c9dcc2d39 Fix regression in sparkline style in web UI (#13753)

5.  bundle install
mastodon@cztwitter:~/live$ bundle install
Fetching gem metadata from https://rubygems.org/..........
Your bundle is locked to mimemagic (0.3.5), but that version could not be found in any of the sources listed in your Gemfile. If you haven't changed sources, that means the author of mimemagic (0.3.5) has removed it. You'll need to update your bundle to a version other than mimemagic (0.3.5) that hasn't been removed in order to install.

6. yarn install (as recommended in Upgrade notes)
mastodon@cztwitter:~/live$ yarn install
yarn install v1.22.4
[1/6] Validating package.json...
[2/6] Resolving packages...
[3/6] Fetching packages...
info fsevents@2.1.3: The platform "linux" is incompatible with this module.
info "fsevents@2.1.3" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.13: The platform "linux" is incompatible with this module.
info "fsevents@1.2.13" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/6] Linking dependencies...
warning " > react-redux-loading-bar@4.0.8" has incorrect peer dependency "react-redux@^3.0.0 || ^4.0.0 || ^5.0.0".
[5/6] Building fresh packages...
[6/6] Cleaning modules...
warning Your current version of Yarn is out of date. The latest version is "1.22.19", while you're on "1.22.4".
info To upgrade, run the following command:
$ sudo apt-get update && sudo apt-get install yarn
Done in 31.36s.

7. RAILS_ENV=production bundle exec rails db:migrate
mastodon@cztwitter:~/live$ RAILS_ENV=production bundle exec rails db:migrate
Could not find loofah-2.5.0 in any of the sources
Run `bundle install` to install missing gems.

Please, could you be so kind and help me?

Thank you and have a nice day. 

### Expected behaviour

no errors appear

### Actual behaviour

bundle is locked by mimemagic 0.3.5

### Detailed description

I created droplet from Mastodon image in Digital Ocean. There is installed 3.1.3 version of Mastodon.

I followed "upgrading to a new release" guide in https://docs.joinmastodon.org/admin/upgrading/ and Upgrade notes for 3.1.4 version. I got mimemagic error and I cannot continue.

After "bundle install" command I got error:
Your bundle is locked to mimemagic (0.3.5), but that version could not be found in any of the sources listed in your Gemfile. If you haven't changed sources, that means the author of mimemagic (0.3.5) has removed it. You'll need to update your bundle to a version other than mimemagic (0.3.5) that hasn't been removed in order to install.

I continued by next step - yarn install, I got another error:
info fsevents@2.1.3: The platform "linux" is incompatible with this module.
info "fsevents@2.1.3" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.13: The platform "linux" is incompatible with this module.
info "fsevents@1.2.13" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/6] Linking dependencies...
warning " > react-redux-loading-bar@4.0.8" has incorrect peer dependency "react-redux@^3.0.0 || ^4.0.0 || ^5.0.0".
[5/6] Building fresh packages...
[6/6] Cleaning modules...
warning Your current version of Yarn is out of date. The latest version is "1.22.19", while you're on "1.22.4".
info To upgrade, run the following command:
$ sudo apt-get update && sudo apt-get install yarn
Done in 31.36s.

I continues by next steps - RAILS_ENV=production bundle exec rails db:migrate, I got another error:
Could not find loofah-2.5.0 in any of the sources
Run `bundle install` to install missing gems.

When I run "bundle install" command, I was back in the beginning with error bundle locked by mimemagic.

Please, could you be so kind and help me?

Thank you and have a nice day. 

### Specifications

Mastodon 3.1.3

**Comment by ClearlyClaire** [comment]
Due to dependency issues, some past Mastodon releases are unfortunately uninstallable. Try updating directly to v3.5.3, but please take a note of individual update instructions.

**Comment by olivievranska** [comment]
> Due to dependency issues, some past Mastodon releases are unfortunately uninstallable. Try updating directly to v3.5.3, but please take a note of individual update instructions.

Hi Claire,

you mean at first follow Upgrade notes for v3.5.2 and then follow Upgrade notes for 3.5.3?

i apologize my stupid questions, I'm not so familiar with git builds...

Thank you very much!

**Comment by ClearlyClaire** [comment]
From what I understand, you are using 3.1.3 and attempted to migrate to 3.1.4, but I think you should try migrating directly to 3.5.3, so starting with `git checkout v3.5.3`.

When I said you should take a note of individual update instructions, I meant there are sometimes special instructions when upgrading from one version to another, so you should read instructions from 3.1.4, 3.1.5, 3.2.0 and so on to make sure you are not missing any step. Anyway, I have done so, and the steps are basically:

1. `git checkout v3.5.3`
2. `bundle install`
3. `yarn install`
4. `SKIP_POST_DEPLOYMENT_MIGRATIONS=true RAILS_ENV=production bundle exec rails db:migrate` (require by a bunch of versions, including https://github.com/mastodon/mastodon/releases/tag/v3.3.0)
5. `RAILS_ENV=production bundle exec rails assets:precompile`
6. restart services (`systemctl reload mastodon-web && systemctl restart mastodon-sidekiq mastodon-streaming`)
7. `RAILS_ENV=production bin/tootctl cache clear` (required by https://github.com/mastodon/mastodon/releases/tag/v3.3.0 and https://github.com/mastodon/mastodon/releases/tag/v3.4.0)
8. `RAILS_ENV=production bundle exec rails db:migrate`

Also make sure your system dependencies are up to date (see https://github.com/mastodon/mastodon/releases/tag/v3.5.3 for current dependencies)

**Comment by beep** [comment]
Just wanted to quietly add a note of thanks, @ClearlyClaire: [your comment](https://github.com/mastodon/mastodon/issues/20128#issuecomment-1307288407) was massively helpful, and unblocked me on an upgrade I was struggling with this morning. Thank you so much!

**Comment by olivievranska** [comment]
Yes, thank you very much, @ClearlyClaire! You saved hours of my frustration.

Good news for all - on Digital Ocean is now image with latest Mastodon 3.5.3, so these steps are not necessary for old Mastodon droplet owners.

I will only need to test migration on snapshopts :-))

**Comment by tmonkey** [comment]
> From what I understand, you are using 3.1.3 and attempted to migrate to 3.1.4, but I think you should try migrating directly to 3.5.3, so starting with `git checkout v3.5.3`.
> 
> When I said you should take a note of individual update instructions, I meant there are sometimes special instructions when upgrading from one version to another, so you should read instructions from 3.1.4, 3.1.5, 3.2.0 and so on to make sure you are not missing any step. Anyway, I have done so, and the steps are basically:
> 
> 1. `git checkout v3.5.3`
> 2. `bundle install`
> 3. `yarn install`
> 4. `SKIP_POST_DEPLOYMENT_MIGRATIONS=true RAILS_ENV=production bundle exec rails db:migrate` (require by a bunch of versions, including https://github.com/mastodon/mastodon/releases/tag/v3.3.0)
> 5. `RAILS_ENV=production bundle exec rails assets:precompile`
> 6. restart services (`systemctl reload mastodon-web && systemctl restart mastodon-sidekiq mastodon-streaming`)
> 7. `RAILS_ENV=production bin/tootctl cache clear` (required by https://github.com/mastodon/mastodon/releases/tag/v3.3.0 and https://github.com/mastodon/mastodon/releases/tag/v3.4.0)
> 8. `RAILS_ENV=production bundle exec rails db:migrate`
> 
> Also make sure your system dependencies are up to date (see https://github.com/mastodon/mastodon/releases/tag/v3.5.3 for current dependencies)

Thanks for this. I'm in the same boat, but after I checkout v3.5.3 and do bundle install, I get 
`rbenv: version `3.0.3' is not installed (set by /home/mastodon/live/.ruby-version)`

Do I need to update Ruby?  How do I do that?

**Comment by maxpollack** [comment]
hey @olivievranska sorry to lean on you, but do you have insight into the migration path now that the droplet has been upgraded? trying to figure out how to move my instance over without breaking everything :)

**Comment by olivievranska** [comment]
> hey @olivievranska sorry to lean on you, but do you have insight into the migration path now that the droplet has been upgraded? trying to figure out how to move my instance over without breaking everything :)

Hey @maxpollack not yet, but it's planned to create postgresql and redis db backup of my running instance @cztwitter.cz and then copy /home/mastodon/live/system folder to sandbox. And before databases import I have to find how to separate sandbox from productive instance because I want only test on different IP address. Maybe is sufficient to not touch DNS records but I'm not sure because of Letsencrypt certificate...

**Comment by tmonkey** [comment]
See @ClearlyClaire 's post above. I followed it (upgrading Mastodon to 3.5.3 from the version that was installed in the old Droplet) and it worked like a charm. 
https://github.com/mastodon/mastodon/issues/20128#issuecomment-1307288407

**Comment by simultsop** [comment]
> From what I understand, you are using 3.1.3 and attempted to migrate to 3.1.4, but I think you should try migrating directly to 3.5.3, so starting with `git checkout v3.5.3`.
> 
> When I said you should take a note of individual update instructions, I meant there are sometimes special instructions when upgrading from one version to another, so you should read instructions from 3.1.4, 3.1.5, 3.2.0 and so on to make sure you are not missing any step. Anyway, I have done so, and the steps are basically:
> 
> 1. `git checkout v3.5.3`
> 2. `bundle install`
> 3. `yarn install`
> 4. `SKIP_POST_DEPLOYMENT_MIGRATIONS=true RAILS_ENV=production bundle exec rails db:migrate` (require by a bunch of versions, including https://github.com/mastodon/mastodon/releases/tag/v3.3.0)
> 5. `RAILS_ENV=production bundle exec rails assets:precompile`
> 6. restart services (`systemctl reload mastodon-web && systemctl restart mastodon-sidekiq mastodon-streaming`)
> 7. `RAILS_ENV=production bin/tootctl cache clear` (required by https://github.com/mastodon/mastodon/releases/tag/v3.3.0 and https://github.com/mastodon/mastodon/releases/tag/v3.4.0)
> 8. `RAILS_ENV=production bundle exec rails db:migrate`
> 
> Also make sure your system dependencies are up to date (see https://github.com/mastodon/mastodon/releases/tag/v3.5.3 for current dependencies)

For anyone stalling on rbenv 3.0.3 not found, figure out to upgrade rbenv later on repeat ```bundle install```, and then elevate node to v16 then perform ```yarn install```.

**Comment by tmonkey** [comment]
Thanks all. I followed the steps above and it worked...except I'm seeing a CSS problem on the /explore pages and some admin pages, mostly around the account cards. The images are huge and not being sized properly. I poked around a little and compared to other sites and it looks like I'm not getting a footer.scss or admin.scss.

![image](https://user-images.githubusercontent.com/198644/201531042-c0f7e8f7-6bca-4f23-b51f-19882f77e49f.png)
(This is what I'm NOT getting on my /explore page)

Is this having to do with the fact that I'm storing objects on S3? If so does anyone have any advice on how to fix this?

Thanks!

**Comment by tmonkey** [comment]
> See @ClearlyClaire 's post above. I followed it (upgrading Mastodon to 3.5.3 from the version that was installed in the old Droplet) and it worked like a charm. [#20128 (comment)](https://github.com/mastodon/mastodon/issues/20128#issuecomment-1307288407)

OK not perfect. I just noticed a css problem. See below.

**Comment by ClearlyClaire** [comment]
Did the `RAILS_ENV=production bundle exec rails assets:precompile` succeed?

**Comment by tmonkey** [comment]
Yes. But with some incompatibilities:

<img width="1035" alt="image" src="https://user-images.githubusercontent.com/198644/201533045-5e577570-0fd5-4dda-8ae0-e89eea226af5.png">

**Comment by ClearlyClaire** [comment]
It's saying “Everything's up-to-date. Nothing to do”, so it did not truly run. That may be fine, but sometimes failures annoyingly don't register as such.

To make sure it works, I'd do `RAILS_ENV=production bundle exec rails assets:clobber` and try again.

Though I am really not sure that is the issue. Do you have local changes to the repo?

**Comment by tmonkey** [comment]
OK I did that. And it removed /home/mastodon/live/public/assets
Removed webpack output path directory /home/mastodon/live/public/packs

Then I tried RAILS_ENV=production bundle exec rails assets:precompile again and then it went further this time, looks like it was writing some css files but then it threw an error in cli.js:93:

<img width="1017" alt="image" src="https://user-images.githubusercontent.com/198644/201533757-16e64ec4-2f81-4dda-abcd-d4748bb3afd9.png">

**Comment by tmonkey** [comment]
Should I just try pulling 3.5.3 down again and starting over?

**Comment by tmonkey** [comment]
<img width="991" alt="image" src="https://user-images.githubusercontent.com/198644/201535302-3811b781-99b1-4ba2-a1a9-dc9ede976644.png">

**Comment by ClearlyClaire** [comment]
Hm maybe there's an incompatibility with your node version? Which version of node are you using?

**Comment by tmonkey** [comment]
v12.16.1

**Comment by tmonkey** [comment]
No, I haven't made any changes to those files. Should I just try checking out v3.5.3 again?

**Comment by ClearlyClaire** [comment]
Hm… that should be fine. Yeah you could try checking out v3.5.3 again. Can you check `git status`?

**Comment by tmonkey** [comment]
Ah..
<img width="710" alt="image" src="https://user-images.githubusercontent.com/198644/201535969-8c64b3f6-0fb0-43ac-a8d6-66d5001ab252.png">

**Comment by ClearlyClaire** [comment]
That shouldn't be an issue, but you can safely delete that file.

**Comment by ClearlyClaire** [comment]
It really sounds like it's using an old version of `webpack-merge` for some reason, but normally, this should use the one defined in `yarn.lock` that should not be an issue. I don't understand what could make your install use a different version than the one expected.

**Comment by tmonkey** [comment]
<img width="229" alt="image" src="https://user-images.githubusercontent.com/198644/201557966-1d779ecc-45e4-4274-9c03-8304c2702cbb.png">
Looks like it's using webpack-merge@5.8.0

**Comment by tmonkey** [comment]
I tried reinstalling v3.5.3 and got the same compile error. 

Should I go back to v3.5.1?

**Comment by tmonkey** [comment]
OK problem resolved. Thanks to @dariusk who advised me to update Node to v16. I recompiled and everything worked!

**Comment by tmonkey** [comment]
And thanks @ClearlyClaire for all your help too!

**Comment by dariusk** [comment]
@ClearlyClaire I think the documentation needs to be updated to reflect that a more recent Node is now needed. I ran into compilation issues specifically when going from 3.5.2 to 3.5.3 on Node 12 (hence my advice to @tmonkey )

**Comment by dariusk** [comment]
@janedzumerko it depends on how node was installed in your environment in the first place. You should be able to go to https://nodejs.org/en/download/ and get the right version for you. I'm sorry I can't be more specifically helpful without knowing more about your computing environment

**Comment by tmonkey** [comment]
I used NVM.

Here are some guides:
https://www.hostingadvice.com/how-to/update-node-js-latest-version/

I used https://www.digitalocean.com/community/tutorials/how-to-install-node-js-on-ubuntu-18-04 since I'm on Digital Ocean but this shouldn't really matter.

**Comment by janedzumerko** [comment]
Hey @tmonkey  I am using DigitalOcean also. I am following this steps https://www.digitalocean.com/community/tutorials/how-to-install-node-js-on-ubuntu-18-04 and I am on mastodon@mastodon313onubuntu1804... account, but the issue is when I try to run  `sudo apt update` it is asking for `[sudo] password for mastodon:`. I don't remember setting password for this. Should I switch to root user first?

**Comment by janedzumerko** [comment]
@tmonkey Yes, I am upgrading because I have an older Mastodon instance (3.1.3) since I used the droplet before the upgrade.

**Comment by tmonkey** [comment]
That's what happened to me exactly. So, you need to update node as root. Then login to the mastodon system user with `sudo su - mastodon`, cd to the live directory, and then do the sequence @ClearlyClaire outlined above. And you should be fine after that.

**Comment by janedzumerko** [comment]
Great @tmonkey ! I updated Node version from root. Now executing the steps from Claire. 

I just finished with step 5, but for step 6, should I exit mastodon user and go back to root? Because I see that on the official documentation here https://docs.joinmastodon.org/admin/upgrading/#:~:text=After%20you%20have%20executed%20the%20instructions%20from%20the%20release%20notes%2C%20switch%20back%20to%20root%3A

**Comment by tmonkey** [comment]
I believe so. If you can't it'll tell you that you need to be logged in as root.

**Comment by janedzumerko** [comment]
I can confirm that. So before step 6 you should exit from mastodon user, execute step 6 with root, and then connect again to mastodon user and execute 7 and 8.

**Comment by dariusk** [comment]
@ClearlyClaire Not at my computer so I can't confirm but this comment seems to suggest that there are minimum patch versions required for 12 and 14, or just 16+

https://github.com/mastodon/mastodon/issues/20128#issuecomment-1313772910

**Comment by janedzumerko** [comment]
> > @ClearlyClaire I think the documentation needs to be updated to reflect that a more recent Node is now needed. I ran into compilation issues specifically when going from 3.5.2 to 3.5.3 on Node 12 (hence my advice to @tmonkey )
> 
> Oh, interesting. What's the new requirement? Node 12.22.0?

Expected version "^12.22.0 || ^14.17.0 || >=16.0.0"

I updated to Node v12.22.12, and it worked.
